VERSION-RANGE-SPEC.rst

```diff
@@ -0,0 +1,989 @@
+
+- Comparing versions from two different versioning schemes is an error. Even
+  though there may be some similarities between the ``semver`` version of an npm
+  and the ``deb`` version of its Debian packaging, the way versions are compared
+  specific to each versioning scheme and may be different. Tools should report
+  an error in this case.
+
+- All references to sorting or ordering of version constraints means sorting
+  by version. And sorting by versions always implies using the versioning
+  scheme-specified version comparison and ordering.
+
+
+Some of the known versioning schemes
+----------------------------------------
+
+These are a few known versioning schemes for some common Package URL
+`types` (aka. ``ecosystem``).
+
+- **deb**: Debian and Ubuntu https://www.debian.org/doc/debian-policy/ch-relationships.html
+  The comparators are <<, <=, =, >= and >>.
+
+- **rpm**: RPM distros https://rpm-software-management.github.io/rpm/manual/dependencies.html
+  The a simplified rmpvercmp version comparison routine is used by archlinux Pacman.
+
+- **gem**: Rubygems https://guides.rubygems.org/patterns/#semantic-versioning
+  which is similar to ``node-semver`` for its syntax, but does not use semver
+  versions.
+
+- **npm**: npm uses node-semver which is based on semver with its own range
+  notation https://github.com/npm/node-semver#ranges
+  A similar but different scheme is used by Rust
+  https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html
+  and several other package types may use ``node-semver``-like ranges. But most
+  of these related schemes are not strictly the same as what is implemented in
+  ``node-semver``. For instance PHP ``composer`` may need its own scheme as this
+  is not strictly ``node-semver``.
+
+- **pypi**: Python https://www.python.org/dev/peps/pep-0440/
+
+- **cpan**: Perl https://perlmaven.com/how-to-compare-version-numbers-in-perl-and-for-cpan-modules
+
+- **golang**: Go modules https://golang.org/ref/mod#versions use ``semver`` versions
+  with a specific minimum version resolution algorithm.
+
+- **maven**: Apache Maven supports a math interval notation which is rarely seen
+  in practice http://maven.apache.org/enforcer/enforcer-rules/versionRanges.html
+
+- **nuget**: NuGet https://docs.microsoft.com/en-us/nuget/concepts/package-versioning#version-ranges
+  Note that Apache Maven and NuGet are following a similar approach with a
+  math-derived intervals syntax as in https://en.wikipedia.org/wiki/Interval_(mathematics)
+
+- **gentoo**: Gentoo https://wiki.gentoo.org/wiki/Version_specifier
+
+- **alpine**: Alpine linux https://gitlab.alpinelinux.org/alpine/apk-tools/-/blob/master/src/version.c
+  which is using Gentoo-like conventions.
+
+- **generic**: a generic version comparison algorithm (which will be specified
+  later, likely based on a split on any wholly alpha or wholly numeric segments
+  and dealing with digit and string comparisons, like is done in libversion)
```

Comment:
I suggest to split the generic one into 3 types:
- `datetime`: The value is interpreted as timestamp according to RFC3339 and ISO 8601 (for simplicity sake just allow YYYY-MM-DDThh:mm:ss[.ffffff[Z|[+-]hh:mm]]). This is especially useful if you want to communicate something that resides in the cloud (like SAAS) where it is hard to determine the version of the running software or to communicate production dates.
- `string`: A simple string comparison following the "usual" string sorting rules. (This seems to be currently not really cover - as soon as the string contains a digit it would be separated. 
- `dotint`: Split at `.` into integers, ignore strings. Compare those by groups. That would cover 80% of the versions currently around... It includes `4.2` as well as `8.7.190.182.919`.
- `generic`: Split at a set of delimiters (e.g. `.,;:#+-`). Compare those by groups but interpret well-known strings like `alpha`, `beta` as modifier to the group before.

Reply:
@pombredanne What needs to done to get these things integrated?

Reply:
As discussed - we should remove `generic` and introduce a `unknown`. Reasoning: `generic` sounds like 'a magic default that fixes everything'. However, that is hard for implementers ('What do I need to implement exactly?') as well as consumers of tools ('What exactly was implemented? Why do I get different results with different implementations?'). The `unknown` clearly states that the version semantics are not known - and therefore, it can be computed by machines. Users SHOULD avoid to use it.

Reply:
Also: Here are the issues:
- `intdot`: https://github.com/package-url/purl-spec/issues/166
- `datetime`: https://github.com/package-url/purl-spec/issues/263
- `string`: https://github.com/package-url/purl-spec/issues/265
- `libversion`: https://github.com/package-url/purl-spec/issues/266
- `unknown`: as described in https://github.com/package-url/purl-spec/pull/139#discussion_r1387779682